google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,6 +181,266 @@ class ResumableAsyncStreamingReadWriteRpc {
+                          "goes out of scope.";
+      assert(false);
+    }
+    shutdown.get();
+  }
+
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+      assert(!retry_promise_.has_value());
+      retry_promise_.emplace();
+    }
+    auto retry_policy = retry_factory_();
+    auto backoff_policy = backoff_policy_prototype_->clone();
+    Initialize(std::move(retry_policy), std::move(backoff_policy));
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    // TODO(18suresha) implement
+    return make_ready_future(absl::optional<ResponseType>());
+  }
+
+  future<bool> Write(RequestType const&, grpc::WriteOptions) override {
+    // TODO(18suresha) implement
+    return make_ready_future(false);
+  }
+
+  future<void> Shutdown() override {
+    promise<void> root_promise;
+    future<void> root_future =
+        ConfigureShutdownOrder(root_promise.get_future());
+
+    root_promise.set_value();
+    return root_future;
+  }
+
+ private:
+  using UnderlyingStream =
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>;
+
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  future<void> ConfigureShutdownOrder(future<void> root_future) {
+    {
+      std::unique_lock<std::mutex> lk{mu_};
+      switch (stream_state_) {
+        case State::kShutdown:
+          return make_ready_future();
+        case State::kRetrying:
+          stream_state_ = State::kShutdown;
+          root_future = root_future.then(ChainFuture(
+              retry_promise_->get_future().then([this](future<void>) {
+                std::unique_lock<std::mutex> lk{mu_};
+                CompleteUnsatisfiedOps(Status(), lk);
+              })));
+          break;
+        case State::kInitialized:
+          stream_state_ = State::kShutdown;
+          if (in_progress_read_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_read_->get_future()));
+          }
+          if (in_progress_write_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_write_->get_future()));
+          }
+          std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+              stream = std::move(stream_);
+          CompleteUnsatisfiedOps(Status(), lk);
+          root_future = root_future.then([stream](future<void>) {
+            return future<void>(stream->Finish());
+          });
+      }
+    }
+    return root_future;
+  }
+
+  void SetReadWriteFutures(std::unique_lock<std::mutex>& lk) {
```

Comment:
Optional: I think I would have received the argument by value, and force the caller to reacquire the mutex if they needed to.